src/AbstractActionHandler.ts

```diff
@@ -189,22 +197,64 @@ export abstract class AbstractActionHandler {
    */
   protected async handleActions(
     state: any,
-    block: Block,
     context: any,
+    nextBlock: NextBlock,
     isReplay: boolean,
   ): Promise<void> {
+    const { block } = nextBlock
     const { blockInfo } = block
 
-    const versionedActions = await this.applyUpdaters(state, block, isReplay, context)
+    const versionedActions = await this.applyUpdaters(state, block, context, isReplay)
     if (!isReplay) {
-      this.runEffects(versionedActions, block, context)
+      this.runEffects(versionedActions, context, nextBlock)
     }
 
     await this.updateIndexState(state, block, isReplay, this.handlerVersionName, context)
     this.lastProcessedBlockNumber = blockInfo.blockNumber
     this.lastProcessedBlockHash = blockInfo.blockHash
   }
 
+  private range(start: number, end: number) {
+    return Array(end - start).fill(0).map((_, i: number) => i + start)
+  }
+
+  private runOrDeferEffect(
+    effect: Effect,
+    payload: any,
+    nextBlock: NextBlock,
+    context: any,
+  ) {
+    const { block, lastIrreversibleBlockNumber } = nextBlock
+    if (!effect.deferUntilIrreversible || block.blockInfo.blockNumber <= lastIrreversibleBlockNumber) {
+      effect.run(payload, block, context)
+    } else if (!this.deferredEffects[block.blockInfo.blockNumber]) {
```

Comment:
I think we could simplify this entirely into a single else condition.  The else if isn't really making a decision different from the else, it's just a situation where the array isn't initialized.

```
else {
  if(!this.deferredEffects[block.blockInfo.blockNumber]) {
    !this.deferredEffects[block.blockInfo.blockNumber] = []
  }

  this.deferredEffects[block.blockInfo.blockNumber].push(() => effect.run(payload, block, context)) 
}

Reply:
I think that reads a bit more clearly as to the intent of the block

Reply:
Was going back and forth on this... if only Javascript had a `defaultdict`

Reply:
I propose a motion to rewrite this entire codebase in python.  @jeffreyssmith2nd can you second this.  I figure we could probably do it all in this PR right?